File: search/array-based-search/linear-search/java/LinearSearchTest.java
import org.junit.Test;
import static org.junit.Assert.*;

/**
 * Unit tests for the LinearSearch class.
 */
public class LinearSearchTest {

    @Test
    public void testTargetAtBeginning() {
        int[] arr = {1, 2, 3};
        assertEquals("Should find target at beginning", 0, LinearSearch.linearSearch(arr, 1));
    }

    @Test
    public void testTargetInMiddle() {
        int[] arr = {10, 20, 30, 40};
        assertEquals("Should find target in middle", 2, LinearSearch.linearSearch(arr, 30));
    }

    @Test
    public void testTargetAtEnd() {
        int[] arr = {10, 20, 30};
        assertEquals("Should find target at end", 2, LinearSearch.linearSearch(arr, 30));
    }

    @Test
    public void testTargetNotPresent() {
        int[] arr = {10, 20, 30};
        assertEquals("Should return -1 if target not found", -1, LinearSearch.linearSearch(arr, 40));
    }

    @Test
    public void testEmptyArray() {
        int[] arr = {};
        assertEquals("Should return -1 for empty array", -1, LinearSearch.linearSearch(arr, 10));
    }

    @Test
    public void testSingleElementFound() {
        int[] arr = {5};
        assertEquals("Should find single element", 0, LinearSearch.linearSearch(arr, 5));
    }

    @Test
    public void testSingleElementNotFound() {
        int[] arr = {5};
        assertEquals("Should return -1 for missing element", -1, LinearSearch.linearSearch(arr, 10));
    }

    @Test
    public void testDuplicateElements() {
        int[] arr = {5, 5, 5};
        assertEquals("Should return first occurrence", 0, LinearSearch.linearSearch(arr, 5));
    }
}

File: search/array-based-search/linear-search/cpp/linear_search_test.cpp
#include <gtest/gtest.h>
#include <vector>


int linearSearch(const std::vector<int>& arr, int target);

// Test: Target at the beginning
TEST(LinearSearchTest, TargetAtBeginning) {
    std::vector<int> arr = {1, 2, 3};
    EXPECT_EQ(linearSearch(arr, 1), 0) << "Should find target at index 0";
}

// Test: Target in the middle
TEST(LinearSearchTest, TargetInMiddle) {
    std::vector<int> arr = {10, 20, 30, 40};
    EXPECT_EQ(linearSearch(arr, 30), 2) << "Should find target at index 2";
}

// Test: Target at the end
TEST(LinearSearchTest, TargetAtEnd) {
    std::vector<int> arr = {10, 20, 30};
    EXPECT_EQ(linearSearch(arr, 30), 2) << "Should find target at index 2";
}

// Test: Target not present
TEST(LinearSearchTest, TargetNotFound) {
    std::vector<int> arr = {10, 20, 30};
    EXPECT_EQ(linearSearch(arr, 40), -1) << "Should return -1 for missing target";
}

// Test: Empty array
TEST(LinearSearchTest, EmptyArray) {
    std::vector<int> arr = {};
    EXPECT_EQ(linearSearch(arr, 10), -1) << "Should return -1 for empty array";
}

// Test: Single element (found)
TEST(LinearSearchTest, SingleElementFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(linearSearch(arr, 5), 0) << "Should find single element at index 0";
}

// Test: Single element (not found)
TEST(LinearSearchTest, SingleElementNotFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(linearSearch(arr, 10), -1) << "Should return -1 for missing element";
}

// Test: Duplicate elements (first occurrence)
TEST(LinearSearchTest, DuplicateElements) {
    std::vector<int> arr = {5, 5, 5};
    EXPECT_EQ(linearSearch(arr, 5), 0) << "Should return first occurrence (index 0)";
}

// Entry point for tests (if running standalone)
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}